arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Fold.kt

```diff
@@ -34,7 +33,8 @@ public suspend fun <Error, A, B> Effect<Error, A>.fold(
     callsInPlace(recover, AT_MOST_ONCE)
     callsInPlace(transform, AT_MOST_ONCE)
   }
-  return fold({ invoke() }, { catch(it) }, { recover(it) }, { transform(it) })
+  val a = recover<Error, A>({ invoke(this) }, { return recover(it) }, { return catch(it) })
+  return catch({ transform(a) }) { catch(it) }
```

Comment:
I dislike that this is needed. You'd expect that all we need is the one-liner:
```recover({ transform(invoke()) }, { recover(it) }, { catch(it) })```
but sadly this has slightly different behaviour when transform invokes a captured `raise` call that was captured in block. This is because fold called `Raise.complete` before it executed the `transform`, while the one-liner calls transform before `Raise.complete`, thus resulting in a captured raise that is immediately invoked. This violates the test `EffectSpec.shiftLeakedResultsInRaiseLeakException` which seems to suggest that this behavior is intentional. Thus, we have to use non-local returns to get rid of all our error cases, and then in our happy path we execute the `transform`, keeping in mind that it can `throw`